pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+						values.deltaListenerCancel()
+					}
+					values.deltaListeners, values.deltaListenerCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.SecretType:
+				if values.deltaSecretNonce == "" || values.deltaSecretNonce == nonce {
+					if values.deltaSecretCancel != nil {
+						values.deltaSecretCancel()
+					}
+					values.deltaSecrets, values.deltaSecretCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.RuntimeType:
+				if values.deltaRuntimeNonce == "" || values.deltaRuntimeNonce == nonce {
+					if values.deltaRuntimeCancel != nil {
+						values.deltaRuntimeCancel()
+					}
+					values.deltaRuntimes, values.deltaRuntimeCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			default:
+				typeURL := req.TypeUrl
+				responseNonce, seen := values.deltaNonces[typeURL]
+				if !seen || responseNonce == nonce {
+					// We must signal goroutine termination to prevent a race between the cancel closing the watch
+					// and the producer closing the watch.
+					if terminate, exists := values.deltaTerminations[typeURL]; exists {
+						close(terminate)
+					}
+					if cancel, seen := values.deltaCancellations[typeURL]; seen && cancel != nil {
+						cancel()
+					}
+
+					var watch chan cache.DeltaResponse
+					watch, values.deltaCancellations[typeURL] = s.cache.CreateDeltaWatch(req, &state)
+
+					// a go-routine. Golang does not allow selecting over a dynamic set of channels.
+					terminate := make(chan struct{})
+					values.deltaTerminations[typeURL] = terminate
+					go func() {
+						select {
+						case resp, more := <-watch:
+							if more {
+								values.deltaResponses <- resp
+							} else {
+								// Check again if the watch is cancelled.
+								select {
+								case <-terminate: // do nothing
+								default:
+									// We cannot close the responses channel since it can be closed twice.
+									// Instead we send a fake error response.
+									values.deltaResponses <- deltaErrorResponse
+								}
+							}
+							break
+						case <-terminate:
+							break
+						}
+					}()
+				}
+			}
+		}
+	}
+}
+
 func (s *server) DeltaStreamHandler(str stream.DeltaStream, typeURL string) error {
-	return nil
+	// a channel for receiving incoming delta requests
+	reqCh := make(chan *discovery.DeltaDiscoveryRequest)
+	reqStop := int32(0)
+
+	go func() {
+		for {
+			req, err := str.Recv()
+			if atomic.LoadInt32(&reqStop) != 0 {
+				return
+			}
+			if err != nil {
+				close(reqCh)
+				return
+			}
+
+			reqCh <- req
+		}
+	}()
+
+	err := s.processDelta(str, reqCh, typeURL)
+	atomic.StoreInt32(&reqStop, 1)
+
+	return err
+}
+
+// When we subscribe, we just want to make the cache know we are subscribing to a resource.
+// Providing a name with an empty version is enough to make that happen.
+func (s *server) subscribe(resources []string, sv map[string]string) {
+	for _, resource := range resources {
+		sv[resource] = ""
```

Comment:
This makes me think that on the cache side it does need to leave entries in the map when they are deleted in order to keep the subscription going?

Reply:
I think that's a true statement. We also encounter the issue of `""` being a valid version. I think we might want to do something like you mention below using `*string` with nil values

Reply:
I think in this case we are explicitly letting envoy know that the resource no longer exists. https://github.com/alecholmez/go-control-plane/blob/incremental/pkg/cache/v3/delta_common.go#L91 And if it still needs the resource it should resubscribe. If we never delete resources then `removed_resources ` field of the [DeltaDiscoveryResponse](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/discovery.proto#deltadiscoveryresponse) will never be used, which seems strange. I also think that this probably doesn't matter, because envoy always requests listeners and clusters in wildcard mode and then subscribes to all dependent resources, so as long as LDS and CDS responses are consistent with the rest of the resources produced by the control plane - we should be good.

Reply:
Right, and in the current cache implementation we still signal what needs to be removed and entries still get deleted: https://github.com/envoyproxy/go-control-plane/blob/main/pkg/cache/v3/delta.go#L81 so I think the functionality should be preserved. @snowp do you know if envoy would just resubscribe if it still needs it? I feel like that makes sense so we don't hold on to potentially stale subscriptions

Reply:
I don't think it makes sense for resource additions/removals to affect the subscription state, but I could be wrong here. I think we need a way to communicate that a resource has been subscribed to within the stream but we haven't been able to provide the resource. Similarly if this resource goes away we want the server to be able to communicate this via `removed_resource`, while the stream state reverts to the state it was in when this resource was requested but no such resource existed.

I think it would be good to cover this case in a test, seems like a likely edge case.